### ISSUE TITLE
Intermediate SplitStyle mode

### ISSUE THREAD

**Issue by cquiroz** [scala-js/scala-js#4327]
I have been trying to use `snowpack` with scala.js new split modules in development mode.

This is really meant for development mode where I want to change code in my application and want to see the results as soon as possible in the browser. A tool like `snowpack` is able to detect what changed and only load those

I tried the setting:
```scala
    scalaJSLinkerConfig ~= (_.withModuleSplitStyle(ModuleSplitStyle.SmallestModules)),
```
And that will create modules for each class and indeed only the few that change will be updated and captured by `snowpack`

However for my test application there is a large set of modules, in my case 3172 files including e.g. one for each scala class in use, or  dependencies like `scalajs-react`

```
ls -al target/scala-2.13/seqexec_fastopt/
...
-rw-------     1 cquiroz  staff    11819 Dec 11 14:28 japgolly.scalajs.react.extra.router.RoutingRules.js
...
-rw-------     1 cquiroz  staff    25248 Dec 11 17:25 web.client.table.package$.js
...
-rw-------     1 cquiroz  staff     1190 Dec 11 14:28 java.lang.ThreadLocal.js
...
```

Though `snowpack` only detects the changed files, the sheer amount of files means it needs to check a lot of files and it also puts pressure on the browser as well (needing to track thousands of modules).

What would be great (I have no idea how involved this is) is to have modules grouped, e.g. all from dependency X into an `X.js` and keep the application classes split, this would let the dev tool to load application files as they change, while caching dependencies grouped in "bigger" files

I don't if this is feasible but I'll open this to discuss

**Comment by gzm0** [comment]
Heya, thanks for trying this out and giving us feedback. It's cool to see new stuff being used :)

Just to get a feel of what we are dealing with, could you put ballpark numbers on a bunch of things? (I do not work on a "real" SJS codebase myself so I just wanna calibrate my gut feelings for this). These need not be exact, e.g. ~10s v.s. ~1s is enough.

> And that will create modules for each class and indeed only the few that change will be updated and captured by snowpack

How many modules change usually? How often do classes from dependencies change (global optimization means that a change in your code can cause a dependency's code to change)?

> Though snowpack only detects the changed files, the sheer amount of files means it needs to check a lot of files and it also puts pressure on the browser as well (needing to track thousands of modules).

Do you have ballpark timings of how this behaves in different modes? Especially: Does `SmallestModules` improve matters (by how much), does it make matters worse?

> What would be great (I have no idea how involved this is) is to have modules grouped, e.g. all from dependency X into an X.js and keep the application classes split, this would let the dev tool to load application files as they change, while caching dependencies grouped in "bigger" files

> I don't if this is feasible but I'll open this to discuss

In terms of module analysis, this is relatively easy. There are two points that make this non-trivial:

- When linking, we have lost provenance information of classes: So the linker does not have access to what is a dependency (in sbt/maven's terms) and what is part of the application. This is a solvable problem, but it takes some careful designing.
- I'd like to see some evidence that such bundling will actually improve speed: Because of global optimizations, it *might* be that the code generated for libraries changes often. But it might also not be :) The proposed bundling will only be effective in the latter case.

These two points are why I'd like to understand a bit better what is happening here. If you are willing to do some analysis, I'd be happy to provide a hacked-up version in a branch that supports something like:

```scala
scalaJSLinkerConfig ~= (_.withModuleSplitStyle(ModuleSplitStyle.SmallestApplicationModules(
    applicationPackagePrefixes = List("my.package1", "my.package2"))))
```

**Comment by cquiroz** [comment]
Sure, I'd like to help.

To give it a bit more context, the scenario I'm interested on is during development. I'm now using webpack dev server with `scalajs-bundler` and it lets me modify my code, then do `fastOptJS`d, then bundled with webpack and the browser self updates. In many cases not even a reload is needed and I only need to save files for this to happen.

In general the setup above is usable, though as the application gets larger more and more time is spent on `fastOptJS` and the browser needs to work harder. (In one of my projects the non optimized js is in the order of 15-20m)

A newer pack of tools like snowpack can load esm modules directly into the browser so the process above could in principle be optimized by using `SmallestModules` as `fastLinkJS`. I'm still experimenting with this so it maybe the difference is less than what I think it could be.

My tests shows the combination above is working though given that `SmallestModules` produces so many files it puts strain on `snowpack` and the browser having to handle thousands of modules.  Given that dependencies rarely change, and when they do you'll likely reload anyway, it could make sense to group them.

Now in answer to your questions:
* I'd say in a typical development cycle of UI you change between 1-10 classes, that's for the case where you're retouching the UI. It gets much larger when doing things like model changes.
* `fastLinkJS` of a single module seems faster than `fullOptJS` than the whole but I thought the difference would be larger. Maybe it can be measured.
* The initial load with `SmallestModules` is fairly slow, 10-60secs, which again makes sense as you're serving thousands of modules the browser needs to parse.

**Comment by ramnivas** [comment]
@cquiroz I have no experience with Snowpack, so I could be really off here. But I am wondering if your Snowpack is configured to server http2. That, in theory, should remove any difference between loading many small files vs loading fewer bigger files.

**Comment by cquiroz** [comment]
Yes it does in https and you can. See the difference but still, anyway we should be careful on optimizing for a specific tool.

The js tooling world moves too quickly

**Comment by gzm0** [comment]
Thanks for the background your answers @cquiroz a couple of follow-up questions.

> I'd say in a typical development cycle of UI you change between 1-10 classes, that's for the case where you're retouching the UI. It gets much larger when doing things like model changes.

Is that `.scala` files that change or `.js` files? It is *very* important to understand that there is not a 1-1 mapping of changes. So changing a single `.scala` file could (theoretically) change *all* `.js` files. I'm mostly interested in the effect on `.js` files of a single change in the dev cycle.

> fastLinkJS of a single module seems faster than fullOptJS than the whole but I thought the difference would be larger. Maybe it can be measured.

That's expected: `fullOptJS` is slow. I was wondering about the speed difference of `fastLinkJS` + snowpack between `SplitStyle.FewestModules` and `SplitStyle.SmallestModules`.

> The initial load with SmallestModules is fairly slow, 10-60secs, which again makes sense as you're serving thousands of modules the browser needs to parse.

That's also expected :) Usually when doing incremental stuff, it often comes at a cost of making the initial load a bit slower. Do you have number of how much faster / slower subsequent runs are (i.e. a typical change in the dev cycle)?

**Comment by gzm0** [comment]
Cool! Thanks for these numbers.

> I'm a bit surprised there isn't a big difference on updating the large file than a bunch of small ones

That's expected. The linker internally has already been doing everything incrementally. In fact, we didn't add the tracking of which files need to be updated, so it will always generate all files and then simply don't write the ones that didn't change.

Bunch of questions:

> Fast link js updates 15 js files 

Are all of these "application" JS files?

> Let me know if you need me to do more measurements

Do you have the execution times of the snowpack runs on the changes? (essentially this whole thing is about what we can do to for snowpack to optimize its performance, so we should look at these numbers :)).

**Comment by ramnivas** [comment]
It is surprising that `fastLinkJS` output is 29M vs 17M for `fastOptJS`. Where could this extra code be coming from?

**Comment by gzm0** [comment]
- Imports (`import * as $j_java$002eio$002ePrintStream from "./java.io.PrintStream.js";`)
- Qualified references (`$j_java$002eio$002ePrintStream.$c_Ljava_io_PrintStream` instead of `$c_Ljava_io_PrintStream`)

IIRC, nothing else changes with modules.

**Comment by gzm0** [comment]
@cquiroz, what are the build options you are using for this? Notably how is "`fastLinkJS` v.s. `fastOptJS`" different? (quotes, because `fastLinkJS` can do everything `fastOptJS` does, in fact, the latter only copies files over from `fastLinkJS` for backwards compatibility).

**Comment by cquiroz** [comment]
The only difference is `fastLinkJS` has
`    scalaJSLinkerConfig ~= (_.withModuleSplitStyle(ModuleSplitStyle.SmallestModules)),`

I manually comment that for `fastOptJS`

**Comment by cquiroz** [comment]
About the size. I understand  there is some extra size because you do the imports but it still seems a lot. It is probably  unrelated to this issue but e.g. some files were fairly large even though the original scala code was short.

One particular module was 1.1M for a fairly short inner class. I'm not sure if should expect a correlation though.

**Comment by cquiroz** [comment]
> 
> Bunch of questions:
> 
> > Fast link js updates 15 js files
> 
> Are all of these "application" JS files?
Yes only "application" file change, and it is very much on line on what I'd expect, I change a class and its dependencies are re generated
> 
> > Let me know if you need me to do more measurements
> 
> Do you have the execution times of the snowpack runs on the changes? (essentially this whole thing is about what we can do to for snowpack to optimize its performance, so we should look at these numbers :)).

I'll look at this. I'll try to talk to the authors of `snowpack` as well. I think they haven't considered optimizing for thousands of modules

**Comment by gzm0** [comment]
> One particular module was 1.1M for a fairly short inner class. I'm not sure if should expect a correlation though.

You should be able to expect a correlation with the size of that class without splitting. (I'm not sure you can expect a correlation from .scala to .js). If there are any other differences than what I pointed out (imports and qualification), I'd be eager to know.

**Comment by cquiroz** [comment]
On the 1.1M module I see code from other classes. I can't tell if it is inlined there but it seems odd. I prettified the code and put it on a gist.
The module is named after the original: java.time.Clock$SystemClock.js

But I see inside e.g.
https://gist.github.com/cquiroz/1c25e6716a7877154d0612b5704aabd4#file-system-clock-L195
https://gist.github.com/cquiroz/1c25e6716a7877154d0612b5704aabd4#file-system-clock-L15096

Those seems to come from other classes in the project

**Comment by sjrd** [comment]
The module is named after *one* of the classes or contains. It can contain other classes when they have circular dependencies. The other classes will only be in that module, and not replicated in another module.

**Comment by cquiroz** [comment]
@sjrd thanks for the clarification, it makes sense to put them together in case of circular dependencies. 

At any rate the smallest modules mode is really impressive

**Comment by gzm0** [comment]
> At any rate the smallest modules mode is really impressive

TY :) Also took me over half a year to build :P

**Comment by cquiroz** [comment]
A little bit more of benchmarks. I'm only measuring here the initial load with chrome dev tools

## Smallest modules:
* http: 3366 requests, 42.29s
* https with http/2: 3365 requests, 42.69s

## Single modules:
* http: 68 requests, 1.57s
* https with http/2: 68 requests, 1.48s

So there is a large difference as expected, a surprise though is that http/2 doesn't make a real difference

I guess I should try to see the effect of changes but I'm still trying to think how to do it

**Comment by gzm0** [comment]
That `SmallestModules` takes longer to load in a browser is expected. The mode is really only intended for tools that can process things incrementally (a browser might count as such, with caching and all).

**Comment by gzm0** [comment]
Quickly trying to sync on this: My state is that I'd still like to see some performance numbers for incremental (i.e. non-initial) snowpack runs with `SmallestModules` (maybe compared to `FewestModules` or not :)).

**Comment by cquiroz** [comment]
Right, I'll try to get a some numbers, I'm having the feeling that is more on the tooling side than on scala.js but let's measure before jumping into conclusions

**Comment by cquiroz** [comment]
I'm doing tests of this same setup using `vite2` and unlike `snowpack` it handles a large amount of small files (as produced in `SmallestModules`) with grace.

In this context having a separate mode wouldn't be necessary though it ties you to this tool. One could think other tools may catch up in the future though

In case you're interested this is the example I'm testing here's the PR

https://github.com/cquiroz/scalajs-react-semantic-ui/pull/239

**Comment by gzm0** [comment]
Does this mean we can close this for now? Or are you still interested in improving performance with snowpack?

**Comment by cquiroz** [comment]
I've been using `vite` for a month with a large project using `SmallestModules` in development and `FewestModules` in production.  It is working great as it is, all modern browsers can run it in production mode.

In dev mode Chrome does a very good job at handling the than 5k modules produced in `SmallestModules` mode tough of course doing reloads is a bit slow (around 20-30secs). FF will choke after a while due to the amount of modules (I think)

Most of these modules come from dependencies thus rarely change. In that sense having an intermediate mode would be useful reducing the startup time and the work browser have to do. Still it is not a showstopper

Another aspect is that small changes during development will result in n files changed, Often just a few (1-10s). `vite` will detect those and do n reloads. It would be great (though I'm not sure if technically feasible) that these changes would be atomically written to the drive in a way that the "bundler" would only do one reload.

In summary, the current system is workable and brings several benefits over `scalajs-bundler`, I want to test having several modules exported next.

**Comment by cquiroz** [comment]
I wanted to raise this one again. We are hitting the limits of what is manageable with 6k modules on split mode where over 5k come from around 15 dependencies. It would be really nice to be able to convert those 5k files into just one or a few per dependencies while maintaining them split for the current project code

I guess this falls into the nice but not critical features but it could be very useful for UI development with scala.js

**Comment by gzm0** [comment]
Thanks for the update.

Unfortunately, I don't think I'll have the necessary cycles to work on this anytime soon. However, I'm happy to:

- Work out a detailed interface specification (this feature potentially needs collaboration between the linker and the build tool, but there are other options as well).
- Guide a contributor to build it (expect this to be messy, in the best case, this will touch half of the entire linking pipeline, with high quality expectations and a rather evolved codebase).

I'll keep thinking about this, but unfortunately that's all I can realistically offer.

Note to self / maintainers:

We could do this with minimal interface changes, if we base the grouping on the `IRContainer` (essentially: all the Jars together and all the individual files together).

That would:
- Probably do "the right thing" automatically for most projects
  - dump external stuff into big files, stuff in the current sbt build into small files.
  - JARs are always versioned as a whole, so if a Jar changes, a lot of stuff changes anyways, so likely few unnecessary invalidations.
- Force us to move the `IRCache` and the `Linker` closer together, but that would align with goals in #4087 as well (using an `IRCache` and `Linker` of different versions is probably a bad idea anyways).
- Avoid a complicated interface with the build tool.

**Comment by sjrd** [comment]
Instead of binding the linking pipeline to the notion of "dependencies", jars, etc., I would *much* rather base the decomposition on package names (possibly with regular expressions for matching). Something ahead of the linker could build up a list automatically, or it would be left to do manually.

**Comment by cquiroz** [comment]
@gzm0 Thanks a lot, a guide and help  for a contributor would be greatly appreciated. I'll try to look for somebody on our team to contribute this.

@sjrd The package based approach would also be helpful as there are way fewer packages than classes. It may also make it easier to  handle cases like the same package being published by two dependencies

**Comment by gzm0** [comment]
I have been thinking about this a bit.

# Interface

I think at this point I agree we should simply base this on package prefixes on an opt-in basis. Essentially something like:

```
final case class SmallModulesFor(packages: List[String]) extends ModuleSplitStyle
```

Everything in the packages would be split as small as possible, everything else as coarse as possible.

IMO this is simple to:
- Reason about
- Use (notably, due to opt-in, adding new library dependencies will not require reconfig)
- Implement
- Simple to use / reason about

# Implementation

To achieve proper partial splitting in the presence of multiple entry points and / or dynamic dependencies, I believe the following approach may be viable:

1. Split into coarse modules (essentially `FewestModules`).
2. Split coarse modules that contain classes in target packages fully (with `SmallestModules`)
3. Merge small modules that do not contain target classes.

Notes:
- For step 2 to be correct, we require that no strongly connected component spans two coarse modules. Since coarse analysis works by tag propagation along the dependency edges, we can easily see that all classes in the same strongly connected component must have the same tags. Per consequence, no strongly connected component can span over two coarse modules.
- If feasible, step 2 / 3 are done in a single pass.
- Step 3 is non-trivial: At first I thought that simply all components not containing a target class can be merged. However, this is not the case if we have `a -> b -> c` where `b` is a split target, but `a, c` are not.
- We could disallow dependencies from non-split targets to split targets: This would likely not affect users much, but has deeper cutting implications on coupling between the Analyzer (where we would have to detect this condition) and the module splitter.

**Comment by sjrd** [comment]
That interface does look pretty good to me. :)

> We could disallow dependencies from non-split targets to split targets: This would likely not affect users much, but has deeper cutting implications on coupling between the Analyzer (where we would have to detect this condition) and the module splitter.

I don't think this is a viable restriction. Even if it is true in the source program, it may be violated by optimizer transformations.

**Comment by gzm0** [comment]
> I don't think this is a viable restriction. Even if it is true in the source program, it may be violated by optimizer transformations.

Ah, yes. Good point. In this case we'll need to build the full-blown analyzer right away. Essentially, we need to be able to group things as follows (where c is a split target):

![graphviz](https://user-images.githubusercontent.com/1140323/156917399-ca82f412-4e86-464b-b0ab-c973cac910ce.svg)

([editable](https://bit.ly/3sJkguf))

**Comment by gzm0** [comment]
The following example illustrates a nice property: In the coarse analysis, we can treat strongly connected components containing a target class very similar to dynamic edges: They are a part of the dependency graph we *must* split along.

![graphviz (1)](https://user-images.githubusercontent.com/1140323/156930026-00c8712d-5729-4694-bd78-9e80a38adaa5.svg)

([editable](https://bit.ly/35OQ6gi))

This suggests we might be better off doing the opposite:

1. Determine target strongly connected components.
2. Coarse analysis for the remaining things.

**Comment by gzm0** [comment]
FYI, WIP branch: https://github.com/gzm0/scala-js/tree/intermediate-splitter

**Comment by cquiroz** [comment]
I did a quick test an for one of my apps we went from 2401 files to 74.
Also the load time in browser went from some secs to negligible.
Thanks a lot for this feature I think it will be of extremely useful
